### ISSUE TITLE
Saving best metrics based on Custom metrics failing (WARNING:tensorflow:Can save best model only with CUSTOM METRICS available, skipping) 

### ISSUE THREAD

**Issue by ravinderkhatri** [keras-team/keras#15684]
I have defined a callback that runs on the epoch end and calculated the metrics. It is working fine in terms of calculating the desired metrics. Below is the function for reference

```
class Metrics(tf.keras.callbacks.Callback):
    def __init__(self, train_tf_data, val_tf_data, model, CLASSES, logs={}, **kwargs):
        super().__init__(**kwargs)
        self.train_tf_data = train_tf_data
        self.val_tf_data = val_tf_data
        self.model = model
        self.CLASSES = CLASSES
        # for train data
        self.train_f1_after_epoch = 0
        self.train_prec_after_epoch = 0
        self.train_recall_after_epoch = 0
        # for val data
        self.val_f1_after_epoch = 0
        self.val_prec_after_epoch = 0
        self.val_recall_after_epoch = 0

    def on_train_begin(self, logs={}):
        self.train_reports = None
        self.val_reports = None
        self.val_f1_after_epoch = 0

    def on_epoch_end(self, epoch, logs={}):
        # for train data
        self.train_reports = test_model(model=self.model, data=self.train_tf_data, 
                                        CLASSES=self.CLASSES)
        self.train_f1_after_epoch = self.train_reports['f1_score']
        self.train_recall_after_epoch = self.train_reports['recall']
        self.train_prec_after_epoch = self.train_reports['precision']

        # for val data
        self.val_reports = test_model(model=self.model, data=self.val_tf_data, 
                                      CLASSES=self.CLASSES)
        self.val_f1_after_epoch = self.val_reports['f1_score']
        self.val_recall_after_epoch = self.val_reports['recall']
        self.val_prec_after_epoch = self.val_reports['precision']

        # saving train results to log dir
        logs["train_f1_after_epoch"]=self.train_f1_after_epoch
        logs['train_precision_after_epoch'] = self.train_prec_after_epoch
        logs['train_recall_after_epoch'] = self.train_recall_after_epoch
        
        # saving val results to log dir
        logs['val_f1_after_epoch'] = self.val_f1_after_epoch
        logs['val_precision_after_epoch'] = self.val_prec_after_epoch
        logs['val_recall_after_epoch'] = self.val_recall_after_epoch


        print('train_reports_after_epoch', self.train_reports)
        print('val_reports_after_epoch', self.val_reports)

```
** .....Some model code .....**

Using this in callback

```
m1 = tf.keras.metrics.CategoricalAccuracy()
m2 = tf.keras.metrics.Recall()
m3 = tf.keras.metrics.Precision()
m4 = Metrics(train_tf_data=train_data, 
             val_tf_data=test_data, model=model, 
             CLASSES=CLASS_NAMES)
optimizers = [
        tfa.optimizers.AdamW(learning_rate=lr * .001 , weight_decay=wd),
        tfa.optimizers.AdamW(learning_rate=lr, weight_decay=wd)

           ]
optimizers_and_layers = [(optimizers[0], model.layers[0]), (optimizers[1], model.layers[1:])]
    
optimizer = tfa.optimizers.MultiOptimizer(optimizers_and_layers)


model.compile(
    optimizer= optimizer,
    loss = 'categorical_crossentropy',
    metrics=[m1, m2, m3],
    )

checkpoint_cb = tf.keras.callbacks.ModelCheckpoint(filepath=checkpoint_path, 
                                                    monitor = 'val_f1_after_epoch',
                                                    save_best_only=True,
                                                    save_weights_only=True,
                                                    mode='max',
                                                    save_freq='epoch',
                                                    verbose=1)
                                                    
checkpoint_cb._supports_tf_logs = False

```
The issue that I am facing is that it is giving me a warning that says

WARNING:TensorFlow: Can save the best model only with val_f1_after_epoch available, skipping

Upon investigating history I found that metrics is available in the history

```
print(list(history.history.keys()))
['loss',
'categorical_accuracy',
'recall',
'precision',
'val_loss',
'val_categorical_accuracy',
'val_recall',
'val_precision',
'train_f1_after_epoch',
'train_precision_after_epoch',
'train_recall_after_epoch',
'val_f1_after_epoch', #this is the metrics
'val_precision_after_epoch',
'val_recall_after_epoch']


```
I think there is a bug in ModelCheckpoint where is it not looking at the custom metrics and not saving the model.

I am using Keras with Tensorflow 2.7 (Also tried this with Tensorflow 2.5)
```

**Comment by chunduriv** [comment]
@ravinderkhatri, In order to expedite the trouble-shooting process, please provide a standalone code or colab link to reproduce the issue reported here. Thanks!

**Comment by ravinderkhatri** [comment]
Sharing below the complete code to re-produce the problem. 

Below is the link to label_map.pbtxt file
https://drive.google.com/file/d/1Qqk3jtJLwA-h3G3sgjMMtvr5JWmPJGcC/view?usp=sharing

Link to tfrecord training file
https://drive.google.com/file/d/1WSuFYHiYKN5AXlRmtMfNhQ_gmfsnK-BE/view?usp=sharing

Link to tfrecord testing file
https://drive.google.com/file/d/1NWOMQ9QnVA5txXwtWVPgZ4oT0Ul7fLyv/view?usp=sharing

Function to get labels from label_pbtxt file
```
def get_label_from_label_map_pbtxt(label_map_pbtxt_path):
    CLASS_NAMES = []
    with open(label_map_pbtxt_path, "r") as file:
        for line in file:
            if 'name:' in line:
                img_class_name = line.split('name:')[-1].replace("'", "").strip()
                CLASS_NAMES.append(img_class_name)
    return CLASS_NAMES


```
Function to read tfrecord file
```
def read_tfrecord(example):
    read_features = {'image/encoded' : tf.io.FixedLenFeature([], dtype = tf.string),
                    'image/object/class/label': tf.io.FixedLenFeature([], dtype = tf.int64)}

    parsed_s_example = tf.io.parse_single_example(serialized=example,
                                                features=read_features)
    encoded_image = parsed_s_example['image/encoded']
    decoded_image = tf.io.decode_image(encoded_image)
    image_label = parsed_s_example['image/object/class/label']
    return decoded_image, image_label


def read_tfrecord_parallelize(tfrecord_data_name):
    """
    Perform parallel reading of data from TFRecordDataset generated by using example_proto
    For more info see: to_gpt_tfrecord_example and get_data_and_write_gpt_tf_record_format
    It uses read_gpt_tfrecord function to read data.
    Args:
        tfrecord_data_name(str): name of tfrecord dataset to read
    Returns:
        dataset(tensorflow.python.data.ops.dataset_ops.ShuffleDataset): a shuffled dataset read from tfrecord_data_name 
    """
    dataset = tf.data.TFRecordDataset(tfrecord_data_name, num_parallel_reads=AUTO)
    dataset = dataset.with_options(option_no_order)
    dataset = dataset.map(read_tfrecord, num_parallel_calls=AUTO)
    dataset = dataset.shuffle(300)
    return dataset


```
By using the above function dataset can be called out as
```
train_data = read_tfrecord_parallelize(train_tfrecord_file_path)
test_data = read_tfrecord_parallelzie(test_tfrecord_file_path

```

Model building code

```
try:
    tpu = tf.distribute.cluster_resolver.TPUClusterResolver.connect()
    strategy = tf.distribute.TPUStrategy(tpu)
except ValueError:
    strategy = tf.distribute.MirroredStrategy()


print('REPLICAS', strategy.num_replicas_in_sync)


img_dataset_path = 'images'

IMG_HEIGHT = 224
IMG_WIDTH = 224
IMAGE_SIZE = [IMG_HEIGHT, IMG_WIDTH]
IMG_CHANNELS = 3
BATCH_SIZE = 16

label_map_pbtxt_path = 'label_map.pbtxt'
# get  total image class
CLASS_NAMES = get_label_from_label_map_pbtxt(label_map_pbtxt_path)
n_labels = len(CLASS_NAMES)
# read tf record dataset
train_dataset_path = 'train.record'
test_dataset_path = 'test.record'
# changes in tf record dataset
train_data = create_dataset(train_dataset_path, batch_size=16, IMG_WIDTH=224, IMG_HEIGHT=224, n_labels=n_labels)
test_data = create_dataset(test_dataset_path, batch_size=16, IMG_WIDTH=224, IMG_HEIGHT=224, n_labels=n_labels)

N_EPOCHS = 200

step = tf.Variable(0, trainable=False)

initial_learning_rate = 0.001
schedule = tf.keras.optimizers.schedules.ExponentialDecay(
    initial_learning_rate,
    decay_steps=14,
    decay_rate=0.8,
    staircase=True)
lr = 1e-0 * schedule(step)
wd = lambda: 1e-2 * schedule(step)

with strategy.scope():
    pretrained_model = tf.keras.applications.MobileNetV2(
                                                    weights='imagenet',
                                                    include_top=False,
                                                    input_shape=[*IMAGE_SIZE, IMG_CHANNELS])
    pretrained_model.trainable = True #fine tuning
    model = tf.keras.Sequential([
                            tf.keras.layers.Lambda(# Convert image from int[0, 255] to the format expect by this model
                            lambda data:tf.keras.applications.mobilenet.preprocess_input(
                                tf.cast(data, tf.float32)), input_shape=[*IMAGE_SIZE, 3]),
                            pretrained_model,
                            tf.keras.layers.GlobalAveragePooling2D()])
    model.add(tf.keras.layers.Dense(64, name='object_dense',kernel_regularizer=tf.keras.regularizers.l2(l2=0.001)))
    model.add(tf.keras.layers.BatchNormalization(scale=False, center = False))
    model.add(tf.keras.layers.Activation('relu', name='relu_dense_64'))
    model.add(tf.keras.layers.Dropout(rate=0.2, name='dropout_dense_64'))
    model.add(tf.keras.layers.Dense(32, name='object_dense_2',kernel_regularizer=tf.keras.regularizers.l2(l2=0.01)))
    model.add(tf.keras.layers.BatchNormalization(scale=False, center = False))
    model.add(tf.keras.layers.Activation('relu', name='relu_dense_32'))
    model.add(tf.keras.layers.Dropout(rate=0.2, name='dropout_dense_32'))
    model.add(tf.keras.layers.Dense(16, name='object_dense_16', kernel_regularizer=tf.keras.regularizers.l2(l2=0.01)))
    model.add(tf.keras.layers.Dense(len(CLASS_NAMES), activation='softmax', name='object_prob'))
    m1 = tf.keras.metrics.CategoricalAccuracy()
    m2 = tf.keras.metrics.Recall()
    m3 = tf.keras.metrics.Precision()
    m4 = Metrics(train_tf_data=train_data, val_tf_data=test_data, model=model, CLASSES=CLASS_NAMES)




optimizers = [
    tfa.optimizers.AdamW(learning_rate=lr * .001 , weight_decay=wd),
    tfa.optimizers.AdamW(learning_rate=lr, weight_decay=wd)
           ]


optimizers_and_layers = [(optimizers[0], model.layers[0]), (optimizers[1], model.layers[1:])]

optimizer = tfa.optimizers.MultiOptimizer(optimizers_and_layers)

model.compile(
    optimizer= optimizer,
    loss = 'categorical_crossentropy',
    metrics=[m1, m2, m3],
    )
checkpoint_path = os.getcwd() + os.sep + 'keras_model'

checkpoint_cb = tf.keras.callbacks.ModelCheckpoint(filepath=checkpoint_path, 
                                                    monitor = 'val_f1_after_epoch',
                                                    save_best_only=True,
                                                    save_weights_only=True,
                                                    mode='max',
                                                    save_freq='epoch',
                                                    verbose=1)

checkpoint_cb._supports_tf_logs = False

history = model.fit(train_data, validation_data=test_data, epochs=N_EPOCHS, callbacks=[checkpoint_cb])


```

**Comment by ravinderkhatri** [comment]
I have created a Google colab notebook to replicate the error 
https://colab.research.google.com/drive/1-6qcPFv2BrdenRm6ncFUfNxDdquJO4l7?usp=sharing

**Comment by ravinderkhatri** [comment]
I was using  modelcallback before metrics callback. 
When I used metric callback  before the modelcheckpoint callback it is working fine. 

history = model.fit(train_data, validation_data=test_data, epochs=N_EPOCHS, callbacks=[m4, checkpoint_cb])
